OpenRA.Mods.Common/Traits/Attack/AttackGarrisoned.cs

```diff
@@ -164,7 +185,7 @@ public override void DoAttack(Actor self, in Target target)
 				paxFacing[a.Actor].Facing = targetYaw;
 				paxPos[a.Actor].SetCenterPosition(a.Actor, pos + PortOffset(self, port));
 
-				if (!a.CheckFire(a.Actor, facing, target, true))
+				if (!a.CheckFire(a.Actor, facing, target))
 					continue;
 
 				if (a.Info.MuzzleSequence != null)
```

Comment:
It looks to me like this here is supposed to simply call all the `notifyAttacks` traits instead of awkwardly adding them to `Armament`. That would be a much cleaner solution. (Even better would be `Armament` offering an event or listener you can add to, but seeing that this already does the muzzle sequence this shouldn't be a problem, so I wouldn't complicate things unless necessary.)

Reply:
> It looks to me like this here is supposed to simply call all the notifyAttacks traits instead of awkwardly adding them to Armament

that's how it used to work, and it looked so similar to other implementations that I missed the difference and accidentally created #21093. The issue is that CheckFire can fire multiple times per tick, so the armament needs to be the one handling the notifying

Reply:
You mean `DoAttack` runs multiple times? I meant running the notifies between the continue and the next if.

Reply:
no, the `CheckFire` contains a do-while loop that can fire multiple times

Reply:
I see. One would assume that it isn't necessary to invoke the notifiers several times in the same tick, but it is inside a delayed action.